### PR TITLE
Code improvements & cleanup

### PR DESCRIPTION
**Pull request by Martinski4GitHub**
Code improvements when stopping/restarting Entware services as well as Diversion service.
Some code cleanup.

### CODE REVIEW

**Comment by ExtremeFiretop** [comment]
Wow how did I not even notice it was downloading both changelogs for this release?

This is a great catch! Just waking up but will do a quick review and test in a few moments! 😁

**Comment by ExtremeFiretop** [comment]
Okay it's gonna be a busy day for me today, girlfriend has a honey-do list for me today (Honey, can you please do...) but I got some time to review now! :)

File: MerlinAU.sh
Comment (ExtremeFiretop) [inline MerlinAU.sh:80]: Approved

File: MerlinAU.sh
Comment (ExtremeFiretop) [inline MerlinAU.sh:135]: Approved

File: MerlinAU.sh
Comment (ExtremeFiretop) [inline MerlinAU.sh:1831]: Approved.

File: MerlinAU.sh
Comment (ExtremeFiretop) [inline MerlinAU.sh:1856]: Smart call, this would be the heaviest clear and also with the function to stop the services giving it a few seconds to settle isn't a bad idea. Sometimes forget that not everyone has ROG router with all the resources. lol

File: MerlinAU.sh
Comment (ExtremeFiretop) [inline MerlinAU.sh:2874]: Ahhh, I see what you did here.
Yeah fair enough, it is starting to get a bit repetitive, time to cleanup and condense down. Approved.

File: MerlinAU.sh
Comment (ExtremeFiretop) [inline MerlinAU.sh:3235]: Smart call again, approved; I was wondering if you were going to make it based on the build number or the model number. (Considering we only had 2 models still getting the 386 builds) but build number is consistent with what we have previously been using, model numbers are really only for model specific changes at this point (such as the blocks)

**Review by ExtremeFiretop — APPROVED** [review]

**Comment by ExtremeFiretop** [comment]
@Martinski4GitHub Great work buddy, love it, clean and professional.

Any other thoughts or changes you want to do? Or should I pump out a quick update? 
Are we almost ready to do a maintenance release with 1.0.8? (Maybe remove the migrate function to cut down some code that is never triggered anymore, or up the minimum version?) or should we wait until 1.0.9?

**Comment by Martinski4GitHub** [comment]
> 
> Wow how did I not even notice it was downloading both changelogs for this release?
> 
Last weekend, when I talked with my cousin who owns an **RT-AX86S**, he mentioned that the router had successfully updated to the latest RMerlin F/W (3004.388.6_2) and that he had not seen any problems at all, and then the conversation moved to something else, and that was the end of that.

Then when I saw the report in the SNB Forum about an RT-AX86S router failing to update due to low RAM available, I recalled the conversation and went to review the script log file (I have access to his router via OpenVPN since I help him manage his home network & router) and so reported what I saw WRT the RAM info entries in my forum post. But while reviewing the log file I saw this entry:

RT-AX86S_ChangeLog_Error

Which I found "odd" but at the time I didn't quite grasp its significance. Then later in the evening, the "light bulb came on" and realized what it meant. So I went back to check the script log file, unzipped the F/W ZIP file, and found the 2 changelog files, which explained the above log entry, and that led me to the eventual fix.

I suspect that having the 2 changelog files may have been an error for this specific release, but in any case, the bug fix should avoid the problem if it happens again, or if having 2 changelogs is going to be the "new normal" from now on.

**Comment by Martinski4GitHub** [comment]
> Okay it's gonna be a busy day for me today, girlfriend has a honey-do list for me today (Honey, can you please do...) but I got some time to review now! :)
>
Ah yes, I know exactly what you mean. I got my own "Honey-do List" which tends to get longer almost every other month. In fact, I have 3 running lists: the standard "Honey-please-do List", the "Honey-must-do List", and the "Honey-this-is-long-overdue-so-take-care-of-it-today-do List" :>) So yeah, I fully understand your situation :>).

**Comment by ExtremeFiretop** [comment]
> > Wow how did I not even notice it was downloading both changelogs for this release?
> 
> Last weekend, when I talked with my cousin who owns an **RT-AX86S**, he mentioned that the router had successfully updated to the latest RMerlin F/W (3004.388.6_2) and that he had not seen any problems at all, and then the conversation moved to something else, and that was the end of that.
> 
> Then when I saw the report in the SNB Forum about an RT-AX86S router failing to update due to low RAM available, I recalled the conversation and went to review the script log file (I have access to his router via OpenVPN since I help him manage his home network & router) and so reported what I saw WRT the RAM info entries in my forum post. But while reviewing the log file I saw this entry:
> 
> RT-AX86S_ChangeLog_Error
> 
> Which I found "odd" but at the time I didn't quite grasp its significance. Then later in the evening, the "light bulb came on" and realized what it meant. So I went back to check the script log file, unzipped the F/W ZIP file, and found the 2 changelog files, which explained the above log entry, and that led me to the eventual fix.
> 
> I suspect that having the 2 changelog files may have been an error for this specific release, but in any case, the bug fix should avoid the problem if it happens again, or if having 2 changelogs is going to be the "new normal" from now on.

This was truly a great catch, fantastic troubleshooting! MerlinAU for sure 'should' of found the matching version in the changelog so the significance would seem to mean it wasn't checking the right log file?

Funny how these things can sometimes slip through the cracks. I think maybe both of us were so focused on getting a good parsing method for the changelog 'contents' we totally skipped over the fact it may contain multiple changelogs in the zip some day lol! 

I also suspect this was a one time thing for this release, looking back through most releases I don't see this being a common thing. But having this logic will never hurt us, if anything it will catch us some day again when we least expect it.

Tell your cousin many thanks and again your the best! I didn't even notice haha!

Do let me know how your feeling about the next release or you have anything else in mind to add :)

**Comment by ExtremeFiretop** [comment]
> > Okay it's gonna be a busy day for me today, girlfriend has a honey-do list for me today (Honey, can you please do...) but I got some time to review now! :)
> 
> Ah yes, I know exactly what you mean. I got my own "Honey-do List" which tends to get longer almost every other month. In fact, I have 3 running lists: the standard "Honey-please-do List", the "Honey-must-do List", and the "Honey-this-is-long-overdue-so-take-care-of-it-today-do List" :>) So yeah, I fully understand your situation :>).

The final list for me seems to be the one that always seems to get longer somehow even though it contains the word "today" 🤔 Questions... Questions... 😁😂

**Comment by Martinski4GitHub** [comment]
> 
> Any other thoughts or changes you want to do? Or should I pump out a quick update? Are we almost ready to do a maintenance release with 1.0.8? (Maybe remove the migrate function to cut down some code that is never triggered anymore, or up the minimum version?) or should we wait until 1.0.9?
>
Yeah, the "migration function" is likely no longer needed for the next version. I'm not sure if changing the required minimum F/W version is needed at this point. I've seen several posts of people going back to previous releases (some have gone several versions back) because of some "instability issues" with the latest F/W. Perhaps at some point later on, those same folks might want to update using our script. Just a thought.

WRT other changes, I was just thinking about whether we should have a toggle switch in the "Advanced Settings" menu to enable/disable creating the backup automatically with BACKUPMON (if installed) every single time. The reason is that many people seem to be doing daily backups anyway and for those folks, doing another backup the same day due to the F/W Update might be something they could avoid if presented as an option. Also, when doing a manual F/W Update to test & check things out, one may want to skip the backup step for some reason.  What do you think? 

It's not a critical change to make for this next release, but something to consider perhaps for the next release.

My 2 cents.

**Comment by ExtremeFiretop** [comment]
> > Any other thoughts or changes you want to do? Or should I pump out a quick update? Are we almost ready to do a maintenance release with 1.0.8? (Maybe remove the migrate function to cut down some code that is never triggered anymore, or up the minimum version?) or should we wait until 1.0.9?
> 
> Yeah, the "migration function" is likely no longer needed for the next version. I'm not sure if changing the required minimum F/W version is needed at this point. I've seen several posts of people going back to previous releases (some have gone several versions back) because of some "instability issues" with the latest F/W. Perhaps at some point later on, those same folks might want to update using our script. Just a thought.
> 
> WRT other changes, I was just thinking about whether we should have a toggle switch in the "Advanced Settings" menu to enable/disable creating the backup automatically with BACKUPMON (if installed) every single time. The reason is that many people seem to be doing daily backups anyway and for those folks, doing another backup the same day due to the F/W Update might be something they could avoid if presented as an option. Also, when doing a manual F/W Update to test & check things out, one may want to skip the backup step for some reason. What do you think?
> 
> It's not a critical change to make for this next release, but something to consider perhaps for the next release.
> 
> My 2 cents.

I'm totally up for it, great idea! 
So maybe let's do this release as is? 

Then for the next release we can remove the migration function and add the new toggle in the advanced options for backupmon if installed. 

We will hold off on increasing the minimum version for now and will reassess later down the router near the next firmware release.

I'll start packaging up this version tomorrow morning for a Sunday release 🙂

**Comment by Martinski4GitHub** [comment]
> 
> I'm totally up for it, great idea! So maybe let's do this release as is?
> 
> Then for the next release we can remove the migration function and add the new toggle in the advanced options for backupmon if installed.
> 
> We will hold off on increasing the minimum version for now and will reassess later down the road near the next firmware release.
> 
> I'll start packaging up this version tomorrow morning for a Sunday release 🙂
>
Yep, sounds good to me. Issue the next release tomorrow with what have today, and postpone any new changes for later.

BTW, based on the backlog & action items I saw last Friday for my day job, I don't think I'll have the time this coming week to work on the toggle switch for BACKUPMON so if you have time & want to do that, go ahead.

**Pull request merged**
